File: location/src/main/java/akhoi/libs/mlct/tools/bcc/ByteConcatReader.java
package akhoi.libs.mlct.tools.bcc;

public class ByteConcatReader {
    private final byte[] content;
    private int position;
    private int posPartial;

    public ByteConcatReader(byte[] content) {
        this.content = content;
    }

    public int readInt(int size) {
        int remaining = Math.clamp(size, 0, 32);
        if (remaining == 0) {
            return 0;
        }

        int remainder = 8 - posPartial;
        int readSize = Math.min(remaining, remainder);
        int result = readByte(0, readSize, remainder);
        remaining -= readSize;
        if (remaining <= 0) {
            return result;
        }

        readSize = Math.min(remaining, 8);
        result = readByte(result, readSize, 8);
        remaining -= readSize;
        if (remaining <= 0) {
            return result;
        }

        readSize = Math.min(remaining, 8);
        result = readByte(result, readSize, 8);
        remaining -= readSize;
        if (remaining <= 0) {
            return result;
        }

        readSize = Math.min(remaining, 8);
        result = readByte(result, readSize, 8);
        remaining -= readSize;
        if (remaining <= 0) {
            return result;
        }

        readSize = Math.min(remaining, 8);
        return readByte(result, readSize, 8);
    }

    public long readLong(int size) {
        if (size <= 32) {
            long value = readInt(size);
            return value & ((1L << size) - 1);
        } else {
            int msbCount = size - 32;
            long msbs = readInt(msbCount) & ((1L << msbCount) - 1);
            long lsbs = readInt(32) & 0xFFFFFFFFL;
            return (msbs << 32) | lsbs;
        }
    }

    public double readDouble(int size) {
        long longValue = readLong(size);
        return Double.longBitsToDouble(longValue);
    }

    public void reset() {
        posPartial = 0;
    }

    public void skip(int count) {
        posPartial += count;
    }

    private int readByte(int value, int size, int remainder) {
        if (position >= content.length) {
            return value;
        }

        int read = ((1 << size) - 1) & (content[position] >>> (remainder - size));
        int posNextByte = posPartial + size;
        position += posNextByte >>> 3;
        posPartial = posNextByte & 7;
        return (value << size) | read;
    }
}

File: location/src/main/java/akhoi/libs/tools/ByteConcatReader.kt
package akhoi.libs.tools

import kotlin.math.min

class ByteConcatReader(private val content: ByteArray) {
    var position: Int = 0

    fun readInt(size: Int): Int {
        if (size <= 0) {
            return 0
        }
        var result = 0
        
        var byte: Int
        var available: Int
        var remaining = min(size, 32)
        val totalBytes = content.size * 8
        while (remaining > 0 && position < totalBytes) {
            byte = position shr 3
            available = 8 - (position and 7)
            val extractBits = min(remaining, available)
            val extracted = ((1 shl extractBits) - 1) and (content[byte].toInt() ushr (available - extractBits))
            result = (result shl extractBits) or extracted
            position += extractBits
            remaining -= extractBits
        }
        return result
    }

    fun readLong(size: Int): Long {
        val value: Long
        if (size <= 32) {
            value = readInt(size).toLong() and (-1L shl size).inv()
        } else {
            val msBitsSize = size - 32
            val msBits = readInt(msBitsSize).toLong() and (-1L shl msBitsSize).inv()
            val lsBits = readInt(32).toLong() and 0xFFFFFFFF
            value = msBits shl 32 or lsBits
        }
        return value
    }

    fun reset() {
        position = 0
    }

    fun align() {
        position = (position + 7) and -8
    }
}